Set up tests that verify we can build a fresh counter app across our Gradle/AGP/Kotlin support range

gmackall opened the pull request:
Sets up tests that verify we can build a fresh counter app across our Gradle/AGP/Kotlin support range.

Post submit only, because the suite takes ~30 minutes to run, and I expect it to be _somewhat_ rare that we break only one of these versions (and therefore it doesn't get caught by existing presubmits).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

mackall-work committed:
set up test suite for gradle/agp/kotlin

mackall-work committed:
fix tests, give proper toString to the tuple

mackall-work committed:
fix tests

mackall-work committed:
fix exisitng android_dependency_version_checking_test.dart

mackall-work committed:
remove comment that is a lie

gmackall committed:
Merge branch 'master' into gradle_agp_test_suite

gmackall commented on packages/flutter_tools/test/integration.shard/android_java_17_dependency_version_range_build_test.dart:
TODO: move these to a different shard, or increase number of shards for the tools test. The shard that this ends up falling on can currently time out with some regularity.

gmackall committed:
Merge branch 'master' into gradle_agp_test_suite

gmackall committed:
Merge branch 'master' into gradle_agp_test_suite

mackall-work committed:
move out of tools_tests

mackall-work committed:
remove dependency on flutter_tools test code

mackall-work committed:
delete unused code

mackall-work committed:
new line

mackall-work committed:
move yet again, this time to devicelab tests

mackall-work committed:
remove from tests.dart

mackall-work committed:
ci.yaml

mackall-work committed:
undo temp commenting

gmackall committed:
Merge branch 'master' into gradle_agp_test_suite

gmackall commented:
Tested with 
```
../../bin/cache/dart-sdk/bin/dart bin/test_runner.dart test -t android_java17_dependency_smoke_tests
```
from
```
//flutter/flutter/dev/devicelab
```

mackall-work committed:
TESTOWNERS

mackall-work committed:
missing import

mackall-work committed:
small fixes

mackall-work committed:
...test.dart -> ...tests.dart

reidbaker commented on .ci.yaml:
```diff
@@ -1215,6 +1215,42 @@ targets:
 
+  - name: Linux android_java11_dependency_smoke_tests
+    recipe: devicelab/devicelab_drone
+    presubmit: false
```
Is this presubmit false because this test is slow?

gmackall replied:
Yes, that was my intention

reidbaker commented on dev/devicelab/bin/tasks/android_java11_dependency_smoke_tests.dart:
```diff
@@ -0,0 +1,28 @@
+import 'package:flutter_devicelab/framework/apk_utils.dart';
+import 'package:flutter_devicelab/framework/framework.dart';
+
+List<VersionTuple> versionTuples = <VersionTuple>[
```
Consider linking to the documentation for where we define the lowest supported versions and where we know that these versions are java 11 compatible.

gmackall replied:
Added a link to the file where we define range, and added a link to compatibility matrix explaining the choice of Java 11

reidbaker commented on dev/devicelab/bin/tasks/android_java17_dependency_smoke_tests.dart:
```diff
@@ -0,0 +1,34 @@
+import 'package:flutter_devicelab/framework/apk_utils.dart';
+import 'package:flutter_devicelab/framework/framework.dart';
+
+// Methodology:
```
Good comment

reidbaker commented on dev/devicelab/lib/framework/apk_utils.dart:
```diff
@@ -29,6 +30,50 @@ final List<String> baseApkFiles = <String> [
 ];
 
+
+const String gradleSettingsFileContent = r'''
```
Please add dart doc.

gmackall replied:
Added

gmackall replied:
Also added dart docs to the main test methods

reidbaker commented on dev/devicelab/lib/framework/apk_utils.dart:
```diff
@@ -477,3 +522,77 @@ String? validateSnapshotDependency(FlutterProject project, String expectedTarget
     ? null : 'Dependency file should have $expectedTarget as target. Instead found $contentSnapshot';
 }
+
+class VersionTuple {
```
Then it is especially important to document that this class is a duplicate and link to/from each implementation so that if improvements or fixes are made they can be applied to both. 

This will be easier if the class is in its own file.

gmackall replied:
Moved to own file + added a comment in both files to suggest mirroring changes.

mackall-work committed:
review

gmackall committed:
Merge branch 'master' into gradle_agp_test_suite

mackall-work committed:
second review piece

mackall-work committed:
more review + missed file

gmackall commented:
Addressed feedback, going to land this now and verify the targets start running correctly and aren't flakey. Will still be marked `bringup` for a while before I unmark. Any reviewers feel free to provide feedback after landing
